infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java

```diff
@@ -676,25 +682,101 @@ protected void doStartMachine(KubernetesServerResolver serverResolver)
     final Map<String, InternalMachineConfig> machineConfigs = environment.getMachines();
     final String workspaceId = getContext().getIdentity().getWorkspaceId();
     LOG.debug("Begin pods creation for workspace '{}'", workspaceId);
+    PodMerger podMerger = new PodMerger();
+    Map<String, Map<String, Pod>> injectablePods = environment.getInjectablePodsCopy();
     for (Pod toCreate : environment.getPodsCopy().values()) {
       ObjectMeta toCreateMeta = toCreate.getMetadata();
-      final Pod createdPod = namespace.deployments().deploy(toCreate);
-      LOG.debug("Creating pod '{}' in workspace '{}'", toCreateMeta.getName(), workspaceId);
-      storeStartingMachine(createdPod, toCreateMeta, machineConfigs, serverResolver);
+      List<PodData> injectables = getAllInjectablePods(toCreate, injectablePods);
+
+      if (injectables.isEmpty()) {
+        final Pod createdPod = namespace.deployments().deploy(toCreate);
+        LOG.debug("Creating pod '{}' in workspace '{}'", toCreateMeta.getName(), workspaceId);
+        storeStartingMachine(createdPod, toCreateMeta, machineConfigs, serverResolver);
+      } else {
+        try {
+          injectables.add(new PodData(toCreate));
+          Deployment merged = podMerger.merge(injectables);
+          merged.getMetadata().setName(toCreate.getMetadata().getName());
+
+          final Pod createdPod = namespace.deployments().deploy(merged);
+          LOG.debug("Creating pod '{}' in workspace '{}'", toCreateMeta.getName(), workspaceId);
+          storeStartingMachine(createdPod, toCreateMeta, machineConfigs, serverResolver);
```

Comment:
may we externalize that from condition (and LOG.debug) because i see the difference only in `createdPod` variable, so seems we can only compute it under `if` condition and `storeStartingMachine` can be common